bridge-stp.in

```diff
@@ -62,18 +83,61 @@ case $2 in
             exit 10
         fi
 
-        # Make sure mstpd is running.
+        # Start mstpd if necessary.
         if ! pidof -c -s mstpd >/dev/null; then
-            echo 'mstpd is not running' >&2
-            exit 3
+            if [ "$MANAGE_MSTPD" != 'y' ]; then
+                echo 'mstpd is not running' >&2
+                exit 3
+            fi
+            echo 'mstpd is not running'
+            echo 'Starting mstpd ...'
+            "$mstpd" $MSTPD_ARGS || exit 3
+
+            # Due to kernel locks, mstpd will not respond to mstpctl until after
+            # this script exits, so `mstpctl addbridge` must be run as an
+            # asynchronous background process.
+            # If `mstpctl addbridge` fails, we could turn STP off and back on
+            # again to fall back to kernel STP mode.  However, that could cause
+            # an infinite loop if mstpd is being started successfully but is
+            # then dying before or when mstpctl connects to it.  To avoid that
+            # possibility, we instead simply turn STP off if `mstpctl addbridge`
+            # fails.
+            ( "$mstpctl" addbridge "$bridge" || brctl stp "$bridge" off ) &
+            exit 0
         fi
 
         # Add bridge to mstpd.
-        exec $bin_dir/mstpctl addbridge $bridge
+        "$mstpctl" addbridge "$bridge" || exit 3
```

Comment:
this line is initially added in commit bdeae079aa38abdabb754a74d81046aec81fccae
maybe move this change there ?